### PR TITLE
Fixed #34750 -- Fixed QuerySet.count() when grouping by unused multi-valued annotations.

### PR DESCRIPTION
**Pull request by felixxm**
Thanks Toan Vuong for the report.

Regression in 59bea9efd2768102fc9d3aedda469502c218e9b7.

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -498,6 +498,9 @@ def get_aggregation(self, using, aggregate_exprs):
                             annotation_mask |= expr.get_refs()
                     for aggregate in aggregates.values():
                         annotation_mask |= aggregate.get_refs()
+                    for annotation_alias, annotation in self.annotation_select.items():
+                        if not getattr(annotation, "contains_aggregate", True):
+                            annotation_mask.add(annotation_alias)
```
Comment (felixxm) [inline django/db/models/sql/query.py:503]: TBH, I'm not convinced :dizzy_face:

Comment (charettes) [reply django/db/models/sql/query.py:503]: @felixxm I think you have the right approach but I'd suggest maybe more explicit usage of `get_group_by_cols`

```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index fed5ce57eb..e7a3d78d2d 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -496,11 +496,17 @@ def get_aggregation(self, using, aggregate_exprs):
                     if isinstance(self.group_by, tuple):
                         for expr in self.group_by:
                             annotation_mask |= expr.get_refs()
+                    elif has_existing_aggregation:
+                        # Avoid eliding expressions that might have an
+                        # incidence on the implicit grouping logic.
+                        for (
+                            annotation_alias,
+                            annotation,
+                        ) in self.annotation_select.items():
+                            if annotation.get_group_by_cols():
+                                annotation_mask.add(annotation_alias)
                     for aggregate in aggregates.values():
                         annotation_mask |= aggregate.get_refs()
-                    for annotation_alias, annotation in self.annotation_select.items():
-                        if not getattr(annotation, "contains_aggregate", True):
-                            annotation_mask.add(annotation_alias)
                     inner_query.set_annotation_mask(annotation_mask)

             # Add aggregates to the outer AggregateQuery. This requires making
```

Comment (charettes) [reply django/db/models/sql/query.py:503]: @felixxm just to make sure, were you able to reproduce with the project provided by the reporter? I see that your adjusted test case avoid the explicit grouping (`values().annotate(Count)`) but the reported case does use explicit grouping.

My proposed suggestion happens to pass against your test but not the reported use case (explicit grouping). I have to admit that I'm a bit surprised that `values("foo").annotate(cnt=Count(), other=F("test"))` doesn't push `"other"` to `self.group_by` like I would have expected.

Comment (felixxm) [reply django/db/models/sql/query.py:503]: > @felixxm just to make sure, were you able to reproduce with the project provided by the reporter?

Yes.

> I see that your adjusted test case avoid the explicit grouping (`values().annotate(Count)`) but the reported case does use explicit grouping.
> 
> My proposed suggestion happens to pass against your test but not the reported use case (explicit grouping). I have to admit that I'm a bit surprised that `values("foo").annotate(cnt=Count(), other=F("test"))` doesn't push `"other"` to `self.group_by` like I would have expected.

I added a second regression test with explicit grouping.

**Review by felixxm — COMMENTED** [review]
@charettes Thanks for the review :+1: I added a release note and changed to `get_group_by_cols()`.

**Comment by felixxm** [comment]
@Toad2186 Does it work for you?

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Review by charettes — APPROVED** [review]
Thanks for the patch @felixxm, looking good to me!

**Comment by felixxm** [comment]
@charettes Thanks for checking :+1:

**Review by nessita — APPROVED** [review]
I'm curious Mariusz why you are not convinced about your solution. While I'm not expert in the ORM, the changes look good.

**Comment by Toad2186** [comment]
> @Toad2186 Does it work for you?

Apologies for the lateness. I've confirmed this works for my (original) query as well.

**Pull request merged**